onnxruntime/core/providers/cpu/nn/qlinearconv.cc

```diff
@@ -228,13 +231,429 @@ Status QLinearConv::Compute(OpKernelContext* context) const {
+  const size_t kernel_dim = group_input_channels * kernel_size;
+
+  const auto* Wdata = static_cast<const uint8_t*>(tensor.DataRaw());
+  W_shape_ = shape;
+
+  auto alloc = Info().GetAllocator(0, OrtMemTypeDefault);
+
+#ifdef MLAS_SUPPORTS_PACKED_GEMM_U8X8
+  packed_W_size_ = MlasGemmPackBSize(group_output_channels, kernel_dim, true);
+
+  if (packed_W_size_ != 0) {
+    auto* packed_W = static_cast<uint8_t*>(alloc->Alloc(SafeInt<size_t>(group_count) * packed_W_size_));
+    packed_W_buffer_ = BufferUniquePtr(packed_W, BufferDeleter(alloc));
+
+    // Allocate a temporary buffer to hold the reordered oihw->ohwi filter for
+    // a single group.
+    //
+    // Note: The size of this buffer is less than or equal to the size of the original
+    // weight tensor, so the allocation size is guaranteed to fit inside size_t.
+    auto* group_reordered_W = static_cast<uint8_t*>(alloc->Alloc(group_output_channels * group_input_channels * kernel_size));
+    BufferUniquePtr group_reordered_W_buffer(group_reordered_W, BufferDeleter(alloc));
+
+    const size_t W_offset = group_output_channels * kernel_dim;
+
+    for (int64_t group_id = 0; group_id < conv_attrs_.group; ++group_id) {
+      ReorderFilter(Wdata, group_reordered_W, group_output_channels, group_input_channels, kernel_size);
+      MlasGemmPackB(group_output_channels, kernel_dim, group_reordered_W, group_output_channels, true, packed_W);
+      packed_W += packed_W_size_;
+      Wdata += W_offset;
+    }
+
+    is_W_packed_ = true;
+    is_packed = true;
+    return Status::OK();
+  }
+#endif
+
+  auto* reordered_W = static_cast<uint8_t*>(alloc->Alloc(SafeInt<size_t>(sizeof(uint8_t)) * shape.Size()));
+  reordered_W_buffer_ = BufferUniquePtr(reordered_W, BufferDeleter(alloc));
+
+  ReorderFilter(Wdata, reordered_W, output_channels, group_input_channels, kernel_size);
+
+  is_W_packed_ = true;
+  is_packed = true;
+  return Status::OK();
+}
+
+Status QLinearConv<int8_t>::Compute(OpKernelContext* context) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  const Tensor* W = is_W_packed_ ? nullptr : context->Input<Tensor>(3);
+  const auto& W_shape = is_W_packed_ ? W_shape_ : W->Shape();
+
+  const int64_t N = X->Shape()[0];
+  const int64_t M = W_shape[0];
+
+  // validate offsets
+  const Tensor* X_zero_point = context->Input<Tensor>(2);
+  const Tensor* W_zero_point = context->Input<Tensor>(5);
+  const Tensor* Y_zero_point = context->Input<Tensor>(7);
+  ORT_ENFORCE(IsScalarOr1ElementVector(X_zero_point),
+              "QLinearConv : input zero point must be a scalar or 1D tensor of size 1");
+  ORT_ENFORCE(IsScalarOr1ElementVector(Y_zero_point),
+              "QLinearConv : result zero point must be a scalar or 1D tensor of size 1");
+
+  auto X_zero_point_value = *(X_zero_point->template Data<uint8_t>());
+  auto Y_zero_point_value = *(Y_zero_point->template Data<uint8_t>());
+
+  const auto& W_zero_point_shape = W_zero_point->Shape();
```

Comment:
>W_zero_point_shape [](start = 14, length = 18)

Looks scale and zero point can be checked in PrePacking.

Reply:
Yeah, but this could be addressed later.